Disable apps based on study metadata

asizemore opened the pull request:
Resolves #731 , child of #667

The two correlation apps require particular data to run. Some studies don't have these data, so for those cases we want to show a disabled app. More specifically,
- correlation assay x metadata requires at least one continuous metadata variable on an entity that is on a one-to-one path with the assay
- correlation assay x assay is specifically for taxa vs functional data. "Functional data" means data from a metagenomic assay, which is an entity that not all studies have. 

I also used the same architecture for the above checks to address #723 

To Dos

- [x] add some isEnabled function to the apps
- [x] disable all vizs in the app if the app is disabled
- [x] can also do this one https://github.com/VEuPathDB/web-monorepo/issues/723 ?
- [x] have the app give a reason that it's not available
- [x] remove all the ts-ignores :)
- [x] repeat for correlation assay v metadata
- [x] improve comments


For testing, some helpful studies include:
- BONUS. Both corrleation apps should be available
- DailyBaby. Should have only the assay x metadata correlation app
- HMP WGS. Should have only the assay x assay (Functional associations) app
- HMP V1-V3. Should have neither app
-

asizemore commented:
Dear reviewers,

Please pay close attention to
1. The logic in `correlationAssayMetadata.tsx` for checking the right variables
2. The typing of `entityTreeToArray` output to ensure that wouldn't break anything else

Thank you!

asizemore commented:
@reviewers i added some studies in the description to help with testing

dmfalke reviewed (commented):
The code looks great. Just a couple of suggestions.

I'm not crazy about the look:
![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/5029423e-2b27-471b-8393-3a218534b64f)

It sort of looks like that app is selected.

What about something like this:

![image](https://github.com/VEuPathDB/web-monorepo/assets/365139/92b22022-1368-4439-8408-b446f23ff6ca)

I removed the background color, and dropped the opacity to `0.4`.

asizemore commented:
> I removed the background color, and dropped the opacity to 0.4.

Ahh this looks much much better. Good idea. I'll see if i can get the "Not applicable..." text to show up a bit more... to be continued!

And @dmfalke  thank you for your other suggestions - they've all been added!

bobular commented on packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayMetadata.tsx:
```diff
@@ -155,3 +161,66 @@ export function CorrelationAssayMetadataConfiguration(
+    (entity) => entity.isManyToOneWithParent
+  );
+  const oneToOneAncestors = ancestorEntities.slice(
+    1, // removing the assay itself
```
So we assume there are no continuous 'metadata variables' on the *same* entity as the assay entity?

asizemore replied:
Right. That assay entity should only have assay data.

bobular reviewed (commented):
The code looks great but no correlation apps are enabled for DailyBaby :disappointed: 

It should have the Correlation "Discover taxa or genes correlated with metadata" app, right? Because it has 1:1:1 PRM:Samples:Assays and there is a continuous Age variable on PRM.

asizemore commented:
@dmfalke what do you think of the below? I faked transparency by lightening the gray of the text to a hue that matches what it'd look like if it was transparent. 
The implementation isn't elegant, though. Check it out in the latest commit!

<img width="914" alt="Screen Shot 2023-12-20 at 5 41 02 AM" src="https://github.com/VEuPathDB/web-monorepo/assets/11710234/1f4338d7-9b71-4909-a0ce-1b13ce756517">

asizemore commented on packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayMetadata.tsx:
```diff
@@ -155,3 +161,66 @@ export function CorrelationAssayMetadataConfiguration(
+  'EUPATH_0000809',
+  'EUPATH_0000813',
+  'EUPATH_0000812',
+];
```
Great question. I just made a [ticket](https://github.com/VEuPathDB/EdaNewIssues/issues/753) for it but i think it could use some more discussion. There wasn't originally a long-term plan because i didn't think that i would end up needing these assay ids at all until i got into the implementation.

To completely remove the hardcoded ids we'd need to add the following annotations:
1. To the study, add `hasMetagenomicData`. I think there's a similar `hasMapData` annotation?
2. To each entity, add `isAssay`. Or i suppose it could be something like `entityType = 'assay'`, if there are other entity types that might be useful to us.

Also open to other solutions!

asizemore commented on packages/libs/eda/src/lib/core/components/computations/plugins/correlationAssayAssay.tsx:
```diff
@@ -198,3 +204,17 @@ export function CorrelationAssayAssayConfiguration(
+  const entities = entityTreeToArray(studyMetadata.rootEntity);
+  const hasMetagenomicData =
+    entities.filter((entity) => entity.id === 'OBI_0002623').length > 0; // OBI_0002623 = Metagenomic sequencing assay
+
```
you're right we should. I was making the assumption that if that entity existed, then it would have a collection. But just becasue that's generally true doesn't mean it will always be the case! I'll add a check

bobular replied:
Also, an `entities.find(...) !== undefined` might be very very slightly marginally more efficient! (because it stops the search as soon as it finds something - but these arrays are tiny, so it will make very little difference)

bobular replied:
Or I guess `entities.some(...)` works too!

bobular replied:
It's still possible that the study with collections and a OBI_0002623 assay with no collections could get through, I think?

Maybe just this?

```
  const hasMetagenomicData = entities.some((entity) => entity.id === 'OBI_0002623' && !!entity.collections?.length);
```

asizemore replied:
ack you're right! All fixed now!

I was interested so I looked into the difference in performance between `some` and `find`. It seems `some` is ever so slightly more efficient, since `find` calls `some`. But you're right it wouldn't matter on an array of this size anyway.

asizemore commented:
> no correlation apps are enabled for DailyBaby 😞

Oh no! Turns out that Age variable is an integer, not a number, so i was missing it. Fixed in the latest commit. Good catch @bobular !

bobular commented:
Looks good to me. I don't think there's another way.

asizemore commented on packages/libs/eda/src/lib/core/components/computations/plugins/abundance.tsx:
```diff
@@ -195,3 +201,20 @@ export function AbundanceConfiguration(props: ComputationConfigProps) {
+  const studyHasCollections = entities.some(
+    (e): e is StudyEntity & Required<Pick<StudyEntity, 'collections'>> =>
+      !!e.collections?.length
+  );
```
True. Simplified in the latest commit!

bobular reviewed (approved):
Looks good to me (though I haven't fired it up again to test).

I left one pedantic logic query.
